### ISSUE TITLE
Variable inheritance in template module doen't work correctly

### ISSUE THREAD

**Issue by psvo** [ansible/ansible#4391]
When using `{% include ... %}`, inherited variables from parent template are not accessible with original names but only prefixed with `l_`. It is different behaviour than plain jinja2 (and is not documented anywhere?).

I created patch with workaround (see the end of this comment), but it still differs from plain jinja2, so I'm not providing pull request.

I don't completely understand purpose of changes made by ansible to jinja2 engine, so I'm not able to provide correct fix, but something like

``` python
# jinja2/runtime.py
 67         for key, value in iteritems(locals):
 68             if key[:2] == 'l_' and value is not missing:
 69                 parent[key[2:]] = value
```

is missing from `_jinja2_vars` applied to `*extras` vars .

**What version of ansible you are using (ansible --version)**

ansible 1.4 (devel 19386c43a7) last updated 2013/10/07 17:38:38 (GMT +200)

**Steps to reproduce the problem**

Run `ansible-playbook -i localhost, -c local test.yml` with following files:

test.yml:

``` yaml
- hosts: localhost
  gather_facts: no
  vars:
    var0: var0_pb
  tasks:
    - template: src=template1.j2 dest=./out1.txt
```

template1.j2:

``` jinja2
{%- set var1 = 'var1_t1' %}
{%- set var2 = 'var2_t1' %}
{%- include 'template2.j2' %}

TEMPLATE1:

var0:       {{ var0|default('undefined') }}
var1:       {{ var1|default('undefined') }}
var2:       {{ var2|default('undefined') }}

l_var0:     {{ l_var0|default('undefined') }}
l_var1:     {{ l_var1|default('undefined') }}
l_var2:     {{ l_var2|default('undefined') }}
```

template2.j2

```
{%- set var2 = 'var2_t2' %}
{%- set l_var2 = 'l_var2_t2' %}

TEMPLATE2:

var0:       {{ var0|default('undefined') }}
var1:       {{ var1|default('undefined') }}
var2:       {{ var2|default('undefined') }}

l_var0:     {{ l_var0|default('undefined') }}
l_var1:     {{ l_var1|default('undefined') }}
l_var2:     {{ l_var2|default('undefined') }}
```

**Expected results**

```
TEMPLATE2:

var0:       var0_pb
var1:       var1_t1
var2:       var2_t2

l_var0:     undefined
l_var1:     undefined
l_var2:     l_var2_t2

TEMPLATE1:

var0:       var0_pb
var1:       var1_t1
var2:       var2_t1

l_var0:     undefined
l_var1:     undefined
l_var2:     undefined
```

Generated with:

``` python
from jinja2 import Environment, FileSystemLoader
env = Environment(loader=FileSystemLoader('.'))
tpl = env.get_template('template1.j2')
print tpl.render(var0='var0_pb')
```

**Actual results**

```
TEMPLATE2:

var0:       var0_pb
var1:       undefined  # should be var1_t1
var2:       var2_t2

l_var0:     var0_pb    # should be undefined
l_var1:     var1_t1    # should be undefined
l_var2:     l_var2_t2

TEMPLATE1:

var0:       var0_pb
var1:       var1_t1
var2:       var2_t1

l_var0:     undefined
l_var1:     undefined
l_var2:     undefined
```

**Results with patch https://github.com/psvo/ansible/commit/e06022e7bede43b6ad4c8c3af0022ea0ea383e12**

```
TEMPLATE2:

var0:       var0_pb
var1:       var1_t1
var2:       var2_t2

l_var0:     var0_pb    # should be undefined
l_var1:     var1_t1    # should be undefined
l_var2:     l_var2_t2

TEMPLATE1:

var0:       var0_pb
var1:       var1_t1
var2:       var2_t1

l_var0:     undefined
l_var1:     undefined
l_var2:     undefined
```

**Comment by mpdehaan** [comment]
This is not really a feature I've seen anyone use. (but not saying it doesn't have value)

If you would like to work on it, please go ahead, but I don't understand how templates would "inherit" anyway.

I think the way to include a snippet of another file would be useful.

I'm not sure what is going on with the patch with the "l_" stuff but I don't think that would be stuff we would want to do to rename variables for sure.

Basically I'm suggesting that folks use Ansible to define variables because it's simpler, "set" will work fine, includes should work fine, but it seems strange to define different variables in different templates when you could install them all at the top?

I'd probably benefit from seeing a more real world (vs abstract) use case to understand what you are trying to do.

**Comment by mpdehaan** [comment]
Incidentally I should point out using the Jinja2 API at a MUCH more basic level is something we want to deal with, in which case I suspect much of this may self-resolve.

There were some earlier patches that were a bit over-complex previously :)

**Comment by psvo** [comment]
Provided example is realy written just to show the problem. I hit this when I tried to use jinja2's for-loops. Something like:

main.j2 (boring cloudformation json stuff stripped):

``` jinja2
{#
   # from ansible inventory
   vpcs:
     vpc01:
       prefix: 10.10
         networks:
           net01:
             net_num: 90
             services:
                ...stripped...
           net02:
             net_num: 91
             services:
               ... stripped...
#}
{% for vpc_key in vpcs %}
  {% set vpc_data = aws.vpcs[vpc_key] %}
  {% include vcp.j2 %}
  {% for net_key in vpc_data.networks %}
    {% set net_data = vpc_data.networks[net_key] %}
    {% include net.j2 %}
  {% endfor %}
{% endfor %}
```

Files vpc.j2 and net.j2 contains some more boring cloudformation json stuff. In this files variables from main.j2 must be prefixed with "l_" to be visible, but should be available directly.

Purpose of this is to have one cloudformation json template to create whole AWS deployment. Software installation on instances is handled via separate playbooks.

But this is going probably little bit offtopic. ;-)

**Comment by mpdehaan** [comment]
Ok thanks for filling that in.

Most likely this will get "fixed" if we can simplify some of our legacy template code, there is some legacy bits in there about setting up contexts that I think may be able to go.

I'll leave this open so we can check when we do, but it may be a while.

**Comment by mpdehaan** [comment]
You might want to use the other ec2 modules vs the cloudformation one as then it's easier to pass variables to them.

**Comment by jctanner** [comment]
@psvo mpdehaan refactored a lot of the templating code recently and I think that might have resolved your issue. 

With latest devel, these are the results from your testcase:

<pre>
jtanner@u1304:~/issues/4391-jinja-inheritance$ cat /tmp/out1.txt 
TEMPLATE2:
var0:       var0_pb
var1:       var1_t1
var2:       var2_t2
l_var0:     undefined
l_var1:     undefined
l_var2:     l_var2_t2
TEMPLATE1:
var0:       var0_pb
var1:       var1_t1
var2:       var2_t1
l_var0:     undefined
l_var1:     undefined
l_var2:     undefined
</pre>

**Comment by jacobweber** [comment]
Did that development change ever get released? I'm using Ansible 1.6, and when I run this test case, I get the OP's "actual results" instead of the "expected results".

In particular, if I set a variable in one template, then include another, the included template doesn't have access to the variable.
